extensions/amp-viralize-player/0.1/amp-viralize-player.js

```diff
@@ -0,0 +1,378 @@
+  }
+
+  /** @override */
+  viewportCallback(visible) {
+    this.dispatchEvent_(VideoEvents.VISIBILITY, {visible});
+  }
+
+  /** @override */
+  pauseCallback() {
+    this.pause();
+  }
+
+  /**
+   * VideoInterface implementations
+   */
+
+  /** @override */
+  supportsPlatform() {
+    return true;
+  }
+
+  /** @override */
+  isInteractive() {
+    return true;
+  }
+
+  /** @override */
+  getCurrentTime() {
+    // Not supported.
+    return 0;
+  }
+
+  /** @override */
+  getDuration() {
+    // Not supported.
+    return 1;
+  }
+
+  /** @override */
+  getPlayedRanges() {
+    // Not supported.
+    return [];
+  }
+
+  /** @override */
+  play(unusedIsAutoplay) {
+    this.VPTCmd_('play');
+  }
+
+  /** @override */
+  pause() {
+    this.VPTCmd_('pause');
+  }
+
+  /** @override */
+  mute() {
+    this.VPTCmd_('mute');
+  }
+
+  /** @override */
+  unmute() {
+    this.VPTCmd_('unmute');
+  }
+
+  /** @override */
+  showControls() {
+    this.VPTCmd_('showControls');
+  }
+
+  /** @override */
+  hideControls() {
+    this.VPTCmd_('hideControls');
+  }
+
+  /** @override */
+  getMetadata() {
+    // Not implemented
+  }
+
+  /** @override */
+  preimplementsAutoFullscreen() {
+    return false;
+  }
+
+  /** @override */
+  preimplementsMediaSessionAPI() {
+    return false;
+  }
+
+  /** @override */
+  fullscreenEnter() {
+    this.VPTCmd_('enterFullscreen');
+  }
+
+  /** @override */
+  fullscreenExit() {
+    this.VPTCmd_('exitFullscreen');
+  }
+
+  /** @override */
+  isFullscreen() {
+    if (this.iframe_) {
+      return isFullscreenElement(this.iframe_);
+    }
+    return false;
+  }
+
+  /** @override */
+  seekTo(unusedTimeSeconds) {
+    // Not implemented
+  }
+
+  /**
+   * Private methods
+   */
+
+  /**
+   * Evaluate the player url
+   * @param {string} zid
+   * @return {string}
+   * @private
+   */
+  getPlayerUrl_(zid) {
+    return addParamsToUrl(`${BASE_URL}/display/?zid=${zid}`, this.extraParams_);
+  }
+
+  /**
+   * Ask player to execute specified command. If player is not yet ready to receive it,
+   * command will be enqueued waiting for player to be loaded
+   *
+   * @param {string} cmd the command to be executed
+   * @private
+   */
+  VPTCmd_(cmd) {
+    if (this.playerReady_) {
+      this.executeVPTCmd_(cmd);
+    } else {
+      this.preloadCmdQueue_.push(cmd);
+    }
+  }
+
+  /**
+   * Send all enqueued commands to the player
+   *
+   * @private
+   */
+  executeCmdQueue_() {
+    for (let i = 0; i < this.preloadCmdQueue_.length; i++) {
+      this.executeVPTCmd_(this.preloadCmdQueue_[i]);
+    }
+    this.preloadCmdQueue_ = [];
+  }
+
+  /**
+   * Immediately send the command to the player. This method assumes that the player is ready
+   * to receive commands
+   *
+   * @param {string} cmd the command to be executed
+   * @private
+   */
+  executeVPTCmd_(cmd) {
+    this.win.vpt.queue.push(() => {
+      this.win.vpt[cmd](this.zid_);
+    });
+  }
```

Comment:
if the method assume the player is ready to to receive commands we can avoid queing the command and directly call the vpt method

Reply:
Agree, I didn't think at that.